File: crackingthecodeinterview/arrays_strings/oneaway.py
from sys import argv

def main():
    argc = len(argv) #len is O(1)
    if(argc > 3 or argc < 2):
        print("This takes two strings")
        return -1

    differencecounter = 0
    firststring       = argv[1]
    firstlength       = len(argv[1])
    secondstring      = argv[2]
    secondlength      = len(argv[2])
    samesize          = firstlength == secondlength
    smallstringlen    = min(firstlength,secondlength) #O(1)
    biggerstringlen   = max(firstlength,secondlength)
    
    if((not samesize) and (smallstringlen + 1 != biggerstringlen)):
        print(firststring + "," + secondstring + " -> " + "false")
        return 0
    elif((not samesize)):
        differencecounter += 1
 
    for i in range(smallstringlen): #O(n)
        if(firststring[i] != secondstring[i]):
            differencecounter += 1

        if(differencecounter > 1):
            print(firststring + "," + secondstring + " -> " + "false")
            return 0

    
    print(firststring + "," + secondstring + " -> " + "true")
    return 0

main()

File: crackingthecodeinterview/arrays_strings/oneaway.cpp
/*
    This program takes two strings and tries to see if there is only one difference between the two
*/
#include <iostream>
#include <string>
#include <algorithm>

using namespace std;

int 
main(int argc, char *argv[])
{
    if(argc < 2 || argc > 3){
        cerr << "This takes two strings" << endl;
        return -1; 
    }

    string firststring = argv[1]; 
    string secondstr   = argv[2];

    int  difference  = 0; 
    int  sizeone     = firststring.length();
    int  sizetwo     = secondstr.length();   
    bool samesize    = (sizeone == sizetwo); 
    int  biggerstr   = max(sizeone,sizetwo);
    int  smallstr    = min(sizeone,sizetwo);
    
    if((not samesize) and (smallstr + 1 != biggerstr)){
        cout << firststring << "," << secondstr << " -> " << "false" << endl;  
        return 0;
    } else if(not samesize){
        difference++;
    }

    for(int i = 0; i < smallstr; i++){
        if(firststring[i] == secondstr[i]){
            difference++; 
            if(difference > 1){
                cout << firststring << "," << secondstr << " -> " << "false" << endl;  
                return 0; 
            }
        }


    }

    cout << firststring << "," << secondstr << " -> " << "true" << endl;  
    return 0;
}
